### PR TITLE
Feature/task manager setfns

### PR DESCRIPTION
**Pull request by RiddhiBagadiaa**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: scheduling/task_run.go [hunk 1]
```diff
@@ -126,25 +125,21 @@ func (t *TaskRunMetadata) Unmarshal(data []byte) error {
...
 	}
 
-	if _, ok := triggerMap["triggerType"]; !ok {
-		return fmt.Errorf("trigger type is missing")
-	}
-
-	switch triggerMap["triggerType"] {
-	case string(oneOffTrigger):
+	switch temp.TriggerType {
+	case oneOffTrigger:
```
Comment (ahmadnazeri) [inline scheduling/task_run.go:129]: I would change this to `OneOffTriggerType` please so it removes the confusion.

File: scheduling/task_run.go [hunk 2]
```diff
@@ -126,25 +125,21 @@ func (t *TaskRunMetadata) Unmarshal(data []byte) error {
...
 		var oneOffTrigger OneOffTrigger
 		if err := json.Unmarshal(temp.Trigger, &oneOffTrigger); err != nil {
 			return fmt.Errorf("failed to deserialize One Off Trigger data: %w", err)
 		}
 		t.Trigger = oneOffTrigger
-	case string(dummyTrigger):
+	case dummyTrigger:
```
Comment (ahmadnazeri) [inline scheduling/task_run.go:135]: similarly, change this to `DummyTriggerType` please

File: scheduling/task_run.go [hunk 3]
```diff
@@ -29,42 +29,41 @@ type Trigger interface {
...
 
 type OneOffTrigger struct {
-	TriggerName string      `json:"triggerName"`
-	TriggerType TriggerType `json:"triggerType"`
+	TriggerName string `json:"triggerName"`
 }
 
 func (t OneOffTrigger) Type() TriggerType {
-	return t.TriggerType
+	return oneOffTrigger
```
Comment (ahmadnazeri) [inline scheduling/task_run.go:36]: return `OneOffTriggerType`

File: scheduling/task_run.go [hunk 4]
```diff
@@ -29,42 +29,41 @@ type Trigger interface {
...
-	TriggerName string      `json:"triggerName"`
-	TriggerType TriggerType `json:"triggerType"`
-	DummyField  bool        `json:"dummyField"`
+	TriggerName string `json:"triggerName"`
+	DummyField  bool   `json:"dummyField"`
 }
 
 func (t DummyTrigger) Type() TriggerType {
-	return t.TriggerType
+	return dummyTrigger
```
Comment (ahmadnazeri) [inline scheduling/task_run.go:49]: return the `DummyTriggerType`

File: scheduling/taskmanager.go [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+			return []TaskRunMetadata{}, fmt.Errorf("failed to unmarshal run record: %v", err)
+		}
+		runs = append(runs, taskRun)
+	}
+	return runs, nil
+}
+
+// Write Methods
+func (t *TaskManager) SetRunStatus(runID TaskRunID, id TaskID, status Status, err error) error {
+	if id == 0 {
```
Comment (ahmadnazeri) [inline scheduling/taskmanager.go:341]: I would change it `<=`

File: scheduling/taskmanager.go [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+		return fmt.Errorf("invalid run id: %d", id)
+	}
+	if status == "" || (status != Success && status != Failed && status != Pending && status != Running) {
```
Comment (ahmadnazeri) [inline scheduling/taskmanager.go:344]: I don't think this is necessary.

File: scheduling/task_run.go [hunk 7]
```diff
@@ -104,7 +104,7 @@ func (t *TaskRunMetadata) Unmarshal(data []byte) error {
 	}
 	t.Name = temp.Name
 
-	if temp.Status == "" || (temp.Status != Success && temp.Status != Failed && temp.Status != Pending) {
+	if temp.Status == "" || (temp.Status != Success && temp.Status != Failed && temp.Status != Pending && temp.Status != Running) {
```
Comment (ahmadnazeri) [inline scheduling/task_run.go:107]: not needed

File: scheduling/taskmanager.go [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+	serializedMetadata, e := metadata.Marshal()
+	if e != nil {
+		return fmt.Errorf("failed to marshal metadata: %v", e)
+	}
+
+	e = t.storage.Set(fmt.Sprintf("tasks/runs/metadata/%d/%s/%d/task_id=%d/run_id=%d", metadata.StartTime.Year(), metadata.StartTime.Month(), metadata.StartTime.Day(), id, metadata.ID), string(serializedMetadata))
+	return e
+}
+
+func (t *TaskManager) SetRunEndTime(runID TaskRunID, id TaskID, time time.Time) error {
```
Comment (ahmadnazeri) [inline scheduling/taskmanager.go:370]: add a check that the end time is after the start time

File: scheduling/taskmanager.go [hunk 9]
```diff
@@ -383,19 +383,55 @@ func (tm *TaskManager) GetAllTaskRuns() (TaskRunList, error) {
 }
 
 // Write Methods
-func (t *TaskManager) SetRunStatus(id TaskRunID, status Status, err error) error {
-	// we will need task id as well
-	return fmt.Errorf("Not implemented")
-}
+func (t *TaskManager) SetRunStatus(runID TaskRunID, id TaskID, status Status, err error) error {
```
Comment (anthonylasso) [inline scheduling/taskmanager.go:386]: to avoid confusion i'd rename `id` to `taskID`

File: scheduling/taskmanager_test.go [hunk 10]
```diff
@@ -536,6 +537,296 @@ func TestGetRunAll(t *testing.T) {
...
+		storage := MemoryStorageProvider{}
+		manager := NewTaskManager(&storage)
+		for _, task := range test.Tasks {
+			_, err := manager.CreateTask(task.Name, task.Type, task.Target)
+			if err != nil && !test.shouldError {
+				t.Fatalf("failed to create task: %v", err)
+			}
+		}
+
+		// var runDefs []TaskRunMetadata
```
Comment (anthonylasso) [inline scheduling/taskmanager_test.go:656]: still need this?

File: scheduling/taskmanager_test.go [hunk 11]
```diff
@@ -536,6 +537,296 @@ func TestGetRunAll(t *testing.T) {
...
+		for _, run := range test.Runs {
+			_, err := manager.CreateTaskRun(run.Name, run.TaskID, run.Trigger)
+			if err != nil && !test.shouldError {
+				t.Fatalf("failed to create task run: %v", err)
+			}
+			// runDefs = append(runDefs, runDef)
```
Comment (anthonylasso) [inline scheduling/taskmanager_test.go:662]: and this?

File: scheduling/taskmanager_test.go [hunk 12]
```diff
@@ -536,6 +537,296 @@ func TestGetRunAll(t *testing.T) {
...
+		storage := MemoryStorageProvider{}
+		manager := NewTaskManager(&storage)
+		for _, task := range test.Tasks {
+			_, err := manager.CreateTask(task.Name, task.Type, task.Target)
+			if err != nil && !test.shouldError {
+				t.Fatalf("failed to create task: %v", err)
+			}
+		}
+
+		// var runDefs []TaskRunMetadata
```
Comment (anthonylasso) [inline scheduling/taskmanager_test.go:793]: comments

**Pull request merged**